lib/puppet/pops/types/p_object_type_extension.rb

```diff
@@ -153,6 +153,11 @@ def simple_name
     @base_type.simple_name
   end
 
+  # @api private
+  def implementation_class(create = true)
+    @base_type.implementation_class(create)
+  end
+
```

Comment:
Would be good to add a test that these types can be serialized.

Does `PAnyType` intentionally doesn't implement the `implementation_class` method so that derived types need to opt-into "yes I can be serialized"?

Reply:
Only the `Object` type implements the `implementation_class` method which means that both the `PObjectType` and the `PObjectTypeExtension` must have it. The latter is just a parameterized variant of the former.
Types like `Any`, `Optional`, `Variant` are abstract. They have no affinity to an implementation and may represent instances of any other type. A serializer will never encounter an instance of an abstract type because no such instance can ever be created.
Basic types like `Float`, `Integer`, `String`, and `Boolean` are hardwired to their implementation. The serializer has hardcoded behavior for those.
Types like `Timestamp` and `Version` can serialize/deserialize themselves as strings (they respond true to the method `roundtrip_with_string`) and will be serialized with a `__pvalue` that is the string representation.
The `Object` type represents a dynamic type that can be implemented without adding a new hardcoded type to the Pcore system. As such, it needs to be associated with an implementation class at runtime. Some instances of `Object` types are mapped to a concrete implementation class (i.e. they are mapped to an existing Ruby class) while others will create the Ruby class on-the-fly the first time it is requested (hence the `create` option).

Reply:
Glad that I tested this Josh! I learned 2 things:
1. PObjectTypeExtensions seem to be converted to data as PObjectTypes unless they're part of another object (like a hash), in which case they're converted as PObjectTypeExtensions
2. There was another method `parameter_info` that needed to be added as well.